### PR TITLE
Make Oximeter producer endpoint registration best effort

### PR DESCRIPTION
**Pull request by jordanhendricks**
The easy fix for #497, moving on if propolis-server couldn't connect to oximeter after a couple of retries. The better fixes are tracked in #513 and oxidecomputer/omicron#3956.

**Testing: Cannot connect to oximeter**
I tested out the error case with the convenient propolis-server `--metric-addr` option (pointing it at an IP/port that isn't running anything), then ensured I could create and run a VM.

Before, this experiment would hang as seen in #497:
```
jordan@rodman:~/src/propolis$ pfexec ./target/release/propolis-server run confs/helios.conf --metric-addr 127.0.0.1:9000 127.0.0.1:8000
Aug 26 12:42:58.921 INFO reservoir too small (0MiB) to use for guest memory
Aug 26 12:42:58.921 INFO Metrics server will use MetricsEndpointConfig { propolis_addr: 127.0.0.1:8000, metric_addr: 127.0.0.1:9000 }
Aug 26 12:42:58.921 INFO Starting server...
Aug 26 12:42:58.921 INFO listening, local_addr: 127.0.0.1:8000
Aug 26 12:43:04.913 INFO accepted connection, remote_addr: 127.0.0.1:42015, local_addr: 127.0.0.1:8000
Aug 26 12:43:04.914 INFO Attempt to register 127.0.0.1:0 with Nexus/Oximeter at 127.0.0.1:9000, uri: /instance, method: PUT, req_id: ad0b9a76-3de3-4e12-8f8b-d211a5131e15, remote_addr: 127.0.0.1:42015, local_addr: 127.0.0.1:8000
Aug 26 12:43:04.960 ERRO Can't connect to oximeter server:
Error registering as metric producer: Communication Error: error sending request for url (http://127.0.0.1:9000/metrics/producers): error trying to connect: tcp connect error: Connection refused (os error 146)
```

With this change, some warnings/errors print in the logs, but then it moves on (the messages are hard to read because of #512):
```
jordan@rodman:~/src/propolis$ pfexec ./target/release/propolis-server run confs/helios.conf --metric-addr 127.0.0.1:9000 127.0.0.1:8000                   
Aug 26 12:17:22.123 INFO reservoir too small (0MiB) to use for guest memory                                                                               
Aug 26 12:17:22.123 INFO Metrics server will use MetricsEndpointConfig { propolis_addr: 127.0.0.1:8000, metric_addr: 127.0.0.1:9000 }                     
Aug 26 12:17:22.123 INFO Starting server...                                                                                                               
Aug 26 12:17:22.123 INFO listening, local_addr: 127.0.0.1:8000                                                                                            
Aug 26 12:17:27.083 INFO accepted connection, remote_addr: 127.0.0.1:61485, local_addr: 127.0.0.1:8000                                                    
Aug 26 12:17:27.083 INFO Attempt to register 127.0.0.1:0 with Nexus/Oximeter at 127.0.0.1:9000, uri: /instance, method: PUT, req_id: a696b50a-a424-4c94-bb
05-12a10f65cece, remote_addr: 127.0.0.1:61485, local_addr: 127.0.0.1:8000                                                                                 
Aug 26 12:17:27.084 INFO listening, local_addr: 127.0.0.1:49861, component: dropshot                                                                      
Aug 26 12:17:27.115 WARN Aug 26 12:17:27.115Could not connect to oximeter (retrying in 1s):                                                               
Error registering as metric producer: Communication Error: error sending request for url (http://127.0.0.1:9000/metrics/producers): error trying to connec
t: tcp connect error: Connection refused (os error 146) , INFOuri :received request to begin graceful shutdown , /instancelocal_addr, :method :127 .0PUT.0
, .1req_id:49861:,  componenta696b50a-a424-4c94-bb05-12a10f65cece:,  remote_addrdropshot:                                                                 
 127.0.0.1:61485, local_addr: 127.0.0.1:8000                                                                                                              
Aug 26 12:17:28.117 INFO listening, local_addr: 127.0.0.1:37062, component: dropshot                                                                      
Aug 26 12:17:28.146 Aug 26 12:17:28.146WARN  INFOCould not connect to oximeter (retrying in 1s):                                                          
Error registering as metric producer: Communication Error: error sending request for url (http://127.0.0.1:9000/metrics/producers): error trying to connec
t: tcp connect error: Connection refused (os error 146) , received request to begin graceful shutdownuri, :local_addr :/instance , 127.method0.:0. 1:PUT37062, , req_idcomponent::  a696b50a-a424-4c94-bb05-12a10f65cecedropshot,                                                                                   
remote_addr: 127.0.0.1:61485, local_addr: 127.0.0.1:8000                                                                                                  
Aug 26 12:17:29.148 ERRO Could not connect to oximeter after 2 retries, uri: /instance, method: PUT, req_id: a696b50a-a424-4c94-bb05-12a10f65cece, remote_
addr: 127.0.0.1:61485, local_addr: 127.0.0.1:8000                                                                                                         
Aug 26 12:17:29.148 WARN Unable to determine Nexus endpoint for IPv4 addresses, uri: /instance, method: PUT, req_id: a696b50a-a424-4c94-bb05-12a10f65cece,
 remote_addr: 127.0.0.1:61485, local_addr: 127.0.0.1:8000   
 Aug 26 12:17:29.148 INFO initializing new VM, bootrom: /home/jordan/src/propolis/ovmf/OVMF_CODE.fd, use_reservoir: false, properties: InstanceProperties {
    id: 6d11c12c-0b89-4c77-a64e-700b6326aea8,                                                                                                             
    name: "vm0",                                                                                                                                          
    description: "propolis-cli generated instance",                                                                                                       
    image_id: 00000000-0000-0000-0000-000000000000,                                                                                                       
    bootrom_id: 00000000-0000-0000-0000-000000000000,                                                                                                     
    memory: 1024,                                                                                                                                         
    vcpus: 4,                                                                                                                                             
}, spec: V0(                                                                                                                                              
    InstanceSpecV0 {                
...
```

### CODE REVIEW

File: bin/propolis-server/src/lib/server.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/server.rs:466]: reminder to file this and update the ref before merging

File: bin/propolis-server/src/lib/stats.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/stats.rs:171]: These look like they should be in the other order.

File: bin/propolis-server/src/lib/stats.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/stats.rs:167]: Should this `warn!()` go above line 157? If it's down here it won't capture the error code from the last call to `Server::start()`.

File: bin/propolis-server/src/lib/stats.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/stats.rs:143]: optional: I would consider rewriting this as `for attempts in 0..N_ATTEMPTS { ... }` and returning `Err` after the loop body. I think that makes it a little easier to reason about when to log the individual "attempt failed" messages and about the number of times the loop will try to connect (is it supposed to be 2 or 3? I think it ends up being 3 in this implementation, but I have to squint at it a bit to figure it out).

**Review by gjcolombo — APPROVED** [review]
A couple things I think might need to be tidied up, but overall LGTM.

File: bin/propolis-server/src/lib/stats.rs
Comment (jordanhendricks) [reply bin/propolis-server/src/lib/stats.rs:143]: I tried it that way initially and was mildly irritated at having to capture the last seen error somehow. I can change it back.

File: bin/propolis-server/src/lib/stats.rs
Comment (jordanhendricks) [reply bin/propolis-server/src/lib/stats.rs:171]: Derp, I totally missed this because the log messages are hard to read (#512).

File: bin/propolis-server/src/lib/stats.rs
Comment (gjcolombo) [reply bin/propolis-server/src/lib/stats.rs:143]: Were you capturing the error just to return it? If the caller's discarding it now anyway (ILTM like it's just calling `.ok()` now), we could just have this routine return an Option instead of a Result.

Comment (jordanhendricks) [reply bin/propolis-server/src/lib/stats.rs:143]: Yeah. We aren't using it in the caller, but it seemed wrong to me to just eat the error all together, since the logic of what to do with that error is a bit higher up. I don't feel that strongly since this isn't the long-term fix we want anyway.

Comment (jordanhendricks) [reply bin/propolis-server/src/lib/stats.rs]: updated to use a for loop and return an Option in 2eb4619

File: bin/propolis-server/src/lib/server.rs
Comment (jordanhendricks) [reply bin/propolis-server/src/lib/server.rs]: done in 2eb4619

File: bin/propolis-server/src/lib/stats.rs
Comment (jordanhendricks) [reply bin/propolis-server/src/lib/stats.rs:156]: should be irrelevant as of 2eb4619

File: bin/propolis-server/src/lib/stats.rs
Comment (jordanhendricks) [reply bin/propolis-server/src/lib/stats.rs]: fixed in b9095d4

**Comment by jordanhendricks** [comment]
Updated the PR description with testing notes for the error case. The only way I know of to test the "happy" path locally is with an omicron deployment, and my cluster has been in a sad state for some time. In order to not block that on debugging my cluster, and given that this change is pretty low-risk, I am going to file a PR to omicron to bump the propolis dep, which should exercise the `end-to-end-tests` set of tests in CI, which includes launching an instance.

**Comment by jordanhendricks** [comment]
CI for the omicron change pulling in this commit passed: oxidecomputer/omicron#3958#511

**Pull request merged**
